File: 0719-find-k-th-smallest-pair-distance/0719-find-k-th-smallest-pair-distance.cpp
class Solution {
public:
    int smallestDistancePair(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());
        int arraySize = nums.size();

        // Largest element in the sorted array
        int maxElement = nums[arraySize - 1];

        // Maximum possible distance
        int maxPossibleDistance = maxElement * 2;

        // Initialize arrays for prefix counts and value counts
        vector<int> prefixCount(maxPossibleDistance, 0);
        unordered_map<int, int> valueCount;

        // Populate the prefixCount array
        int index = 0;
        for (int value = 0; value < maxPossibleDistance; ++value) {
            while (index < arraySize && nums[index] <= value) {
                ++index;
            }
            prefixCount[value] = index;
        }

        // Populate the valueCount map
        for (int i = 0; i < arraySize; ++i) {
            ++valueCount[nums[i]];
        }

        // Binary search for the k-th smallest distance
        int low = 0, high = maxElement;
        while (low < high) {
            int mid = (low + high) / 2;

            // Count pairs with distance <= mid
            int count = countPairs(nums, prefixCount, valueCount, mid);

            // Adjust binary search bounds based on count
            if (count < k) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }
        return low;
    }

private:
    // Count number of pairs with distance <= m
    int countPairs(vector<int>& nums, vector<int>& prefixCount,
                   unordered_map<int, int>& valueCount, int maxDistance) {
        int count = 0;
        int arraySize = nums.size();
        int index = 0;

        while (index < arraySize) {
            int currentValue = nums[index];
            int valueCountForCurrent = valueCount.at(currentValue);

            // Calculate pairs involving current value with distance <=
            // maxDistance
            int pairsWithLargerValues =
                prefixCount[currentValue + maxDistance] -
                prefixCount[currentValue];
            int pairsWithSameValues =
                valueCountForCurrent * (valueCountForCurrent - 1) / 2;

            count += pairsWithLargerValues * valueCountForCurrent +
                     pairsWithSameValues;

            // Skip duplicate values
            while (index + 1 < arraySize && nums[index] == nums[index + 1]) {
                ++index;
            }
            ++index;
        }
        return count;
    }
};

File: 0719-find-k-th-smallest-pair-distance/0719-find-k-th-smallest-pair-distance.py
class Solution:
    def smallestDistancePair(self, nums, k):
        nums.sort()
        array_size = len(nums)

        # Highest element in the sorted array
        max_element = nums[-1]

        # Maximum possible distance
        max_possible_distance = max_element * 2

        # Initialize arrays for prefix counts and value counts
        prefix_count = [0] * max_possible_distance
        value_count = [0] * (max_element + 1)

        # Populate prefix count and value count
        index = 0
        for value in range(max_possible_distance):
            while index < array_size and nums[index] <= value:
                index += 1
            prefix_count[value] = index
        for i in range(array_size):
            value_count[nums[i]] += 1

        # Binary search to find kth smallest distance
        left, right = 0, max_element
        while left < right:
            mid = (left + right) // 2

            # Count pairs with distance <= mid
            count = self._count_pairs(nums, prefix_count, value_count, mid)

            # Adjust binary search bounds based on count
            if count < k:
                left = mid + 1
            else:
                right = mid
        return left

    def _count_pairs(self, nums, prefix_count, value_count, max_distance):
        count = 0
        array_size = len(nums)
        index = 0

        while index < array_size:
            current_value = nums[index]
            value_count_for_current = value_count[current_value]

            # Calculate pairs involving current value with distance <= max_distance
            pairs_with_larger_values = (
                prefix_count[
                    min(current_value + max_distance, len(prefix_count) - 1)
                ]
                - prefix_count[current_value]
            )
            pairs_with_same_values = (
                value_count_for_current * (value_count_for_current - 1) // 2
            )
            count += (
                pairs_with_larger_values * value_count_for_current
                + pairs_with_same_values
            )

            # Skip duplicate values
            while index + 1 < array_size and nums[index] == nums[index + 1]:
                index += 1
            index += 1

        return count